Fixes #44: Puppet issue with java - Bug fix lightblue platform/lightblue puppet#44

luan-cestari opened the pull request:
This PR have enhancement over the constraints and the ordering .It also completed the other alternatives missed to be configured  and added the installation of them in case they are missed.

There is an pending issue that I couldn't put the assign of a variable in the right order (which can really impact us)

Also, it might be desired to use sinle quote all over the paths, even that theses paths doesn't include any space.

Related to issue https://github.com/lightblue-platform/lightblue-puppet/issues/44

luan-cestari commented:
Included stdin so I could use anchor (which is default included on the newest puppet version but I worried about backward comparability)

luan-cestari commented:
Posted about the problem on https://groups.google.com/forum/#!forum/puppet-users (the post needs to be approved) and also on IRC (#puppet channel in FreeNode)

luan-cestari commented:
The link to the post https://groups.google.com/forum/#!topic/puppet-users/CDz_nsjFOKw 

Note: the impact of the line commented is not very high as it is necessary in case the alternatives was not set, which is not the case in the RPM packages provided. If there is no suggestion about how to solve this scenario we can leave another issue open referring the puppet's google group post or even removing the 'alternative install' from that class.

luan-cestari commented:
I will remove the stdlib and make some changes, such as:
- Add a new argument to exemplify the version of java required to be installed (both the jre and sdk)
- Solve the issue around teh dependency between the enviroment to set alternatives. I will use a command like "repoquery  java-1*  | sed 's/i686$//'  | sed 's/x86_64$//'  | sed 's/noarch$//'  | sed 's/.$//' | sed 's/-1:/ -1:/' | sort | uniq" which I will use to create a map that I can get the package version and use that class (such as java-1.8.0-openjdk-1.8.0.25-4.b18.fc21)
- Make the necessary change on the classes that use this package to use an specific java version

dcrissman commented:
I feel like a lot more is being changed in this PR than the subject indicates. Should this maybe be multiple PRs?

luan-cestari commented:
Note: we can add a custom fact trying some distributions of java (such as '''repoquery java-1.8* | fgrep devel | fgrep openjdk | fgrep -v debug | sed 's/i686$//' | sed 's/x86_64$//' | sed 's/noarch$//' | sed 's/.$//' | sed 's/-1:/-/' | sed 's/devel-//'  | sort | uniq | tail -n 1''')

jewzaam commented:
Build failing, looks like lint issues.  See for example https://travis-ci.org/lightblue-platform/lightblue-puppet/jobs/57975437#L400-L403

jewzaam commented on manifests/java.pp:
```diff
@@ -2,50 +2,120 @@
-    ensure => installed,
-    name   => "java-${java_version}-${java_distribution}",
+    ensure => [installed,$java_version],
+    name   => "${java_package_jre_install}",
```
For the package directive ensure should be the version of the RPM, name is the name of the RPM.  For the example of "java-1.7.0-openjdk-devel-1.7.0.71-2.5.3.1.el6.x86_64"

```
package { 'java':
    ensure => '1.7.0.71-2.5.3.1.el6.x86_64',
    name => 'java-1.7.0-openjdk-devel'
}

luan-cestari commented:
Fixed lint problem. Changed the ensure to check the version of the package and I also removed two alternatives configurations that are only present in java 1.8.

I used vagrant to test and it seems to be working fine

dcrissman commented on manifests/java.pp:
```diff
@@ -2,50 +2,108 @@
   package { 'java':
-    ensure => installed,
-    name   => "java-${java_version}-${java_distribution}",
+    ensure => [installed,$java_package_jre_install],
```
You shouldn't need both 'installed' and the version, you can just put the version.

dcrissman commented on manifests/java.pp:
```diff
@@ -2,50 +2,108 @@
   package { 'java-devel':
-    ensure => installed,
-    name   => "java-${java_version}-${java_distribution}-devel",
+    ensure => [installed,$java_package_sdk_install],
```
You shouldn't need both 'installed' and the version, you can just put the version.

dcrissman commented:
I would like to see some spec tests for this.

luan-cestari commented:
@dcrissman , sorry, I thought it had already rspec tests. Fixed
